cmake configuration guidance

Outside this medium we had an exchange where I needed you to set up downloading and inclusion of an open source yaml parser library.   It would be helpful if you could point me and anyone else reading this to an appropriate online source that provides guidance on how to do this in general.  Altrnatively or also, it would be helpful here to at least document the pieces that need to be set up.   From my attempt I know it is at least two things:

1. Changes to the file mspass/CMakeList.txt that defines the overall control of how this is done (I think)
2. To work it this framework a package needs configuration files built that are installed in mspass/cmake.   How those are used and which are required was not at all clear to me from any definitions in the existing files.   

Are there other files that need to be changed/added?

Case in point:   I tried to use yaml-cpp after you set up the configuration, and I get this message when trying to compile MetadataDefinitions using a new yaml parser:
/home/pavlis/src/mspass/cxx/src/lib/utility/MetadataDefinitions.cc:5:10: fatal error: yaml-cpp/yaml.h: No such file or directory
 #include "yaml-cpp/yaml.h"
          ^~~~~~~~~~~~~~~~~
compilation terminated.

cmake did correctly download and build yaml-cpp, but something got messed up in the path set up for include and presumably lib files as well.   i.e.  can see the include file requested here under build, but some path is wrong.   I don't see how that is being set up by cmake.   Can you correct this and let me know what you needed to do to fix it?

Thanks for handling that so quickly.   The example of the fix you created is good education for me on cmake.

Good news is all that works and I've managed to work through all the compile errors I had on my initial attempt at writing a yaml reader for MetadataDefinitions.   The bad news is I'm getting link errors I've not gotten from any other package we have installed by a similar mechanism.   Here is the message I'm getting when mspasspy is trying to link with the new code that uses yaml-cpp:

[ 76%] Linking CXX shared module mspasspy.cpython-36m-x86_64-linux-gnu.so
/usr/bin/ld: ../yaml-cpp/lib/libyaml-cpp.a(scanscalar.cpp.o): relocation R_X86_64_PC32 against symbol `_ZNK4YAML5RegEx14MatchUncheckedINS_16StreamCharSourceEEEiRKT_' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
python/CMakeFiles/mspasspy.dir/build.make:101: recipe for target 'python/mspasspy.cpython-36m-x86_64-linux-gnu.so' failed
make[2]: *** [python/mspasspy.cpython-36m-x86_64-linux-gnu.so] Error 1
CMakeFiles/Makefile2:88: recipe for target 'python/CMakeFiles/mspasspy.dir/all' failed
make[1]: *** [python/CMakeFiles/mspasspy.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2

The suggestion by the compiler to use -fPIC  is not the issue, I think, when I poked around the cmake files and found the link line I think is creating this error:

pavlis-home-linux:~/src/mspass/cxx/build/python/CMakeFiles/mspasspy.dir> cat link.txt
/usr/bin/c++ -fPIC -Wall -O0 -g  -shared  -o mspasspy.cpython-36m-x86_64-linux-gnu.so CMakeFiles/mspasspy.dir/mspass_wrapper.cpp.o ../src/lib/libmspass.a ../src/lib/seismic/libseismic.a ../src/lib/algorithms/libalgorithms.a ../src/lib/utility/libutility.a ../yaml-cpp/lib/libyaml-cpp.a -llapack -lblas 

-fPIC is in that command list, but the issue seems to be that all the libraries referenced are static (.a).   I think we might need to set up the configuration to create a -shared library from at least the yaml-cpp package.   Not sure though.  Any ideas you might have would be appreciated.

I need your help on this one.  I probably could solve this, but I'm 100% sure you can solve this more efficiently than me.   You will recall you found the "dpss" open source package for C++ that you used for multitaper methods.   Figured out (well, maybe not completely - more on that below) how to get cmake to find gsl that was needed for the deconvolution code.   However, when the compilation gets to trying to compile the dpss.cc code I am getting this error:
```
[ 40%] Building CXX object src/lib/deconvolution/CMakeFiles/deconvolution.dir/dpss.cc.o
/Users/pavlis/src/mspass/cxx/src/lib/deconvolution/dpss.cc:39:21: error: use of
      undeclared identifier 'dlamch_'
    double abstol=2*dlamch_(&safe);         //safe minimum absolute tolerance
                    ^
/Users/pavlis/src/mspass/cxx/src/lib/deconvolution/dpss.cc:54:5: error: use of
      undeclared identifier 'dstebz_'
    dstebz_(&range, &order, &n, &vl, &vu, &il, &iu, &abstol, D, E, &m, &...
    ^
/Users/pavlis/src/mspass/cxx/src/lib/deconvolution/dpss.cc:56:9: error: use of
      undeclared identifier 'dstein_'
        dstein_(&n, D, E, &m, eig_val, IBLOCK, ISPLIT, eig_vec, &ldz, WO...
        ^
3 errors generated.
```
Those are lapack fortran functions that were pulled from perf.h when we were using this with antelope.   Not sure if the solution to this is some include from lapack, FC.h, or both.  You cracked the lapack/openblas months ago.  That is why I'm certain you can solve this problem much more efficiently than me.

It appears something is wrong with the new gsl cmake setup.    I found it works fine with MacOS where it somehow found a version of gsl local.  At some point you need to teach me how it does that.  The problem was when I tried to build on my ubuntu machine.  It is downloading the gsl library and compiling it in the build directory.  After it finishes compiling it aborts with this message from cmake:
```
/usr/bin/install -c -m 644 gsl_math.h gsl_pow_int.h gsl_nan.h gsl_machine.h gsl_mode.h gsl_precision.h gsl_types.h gsl_version.h gsl_minmax.h gsl_inline.h '/home/pavlis/src/mspass/cxx/build/gsl/include/gsl'
[ 88%] No test step for 'gsl'
[100%] Completed 'gsl'
[100%] Built target gsl
CMake Error at cmake/gsl.cmake:25 (find_package):
  By not providing "Findgsl.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "gsl", but
  CMake did not find one.

  Could not find a package configuration file provided by "gsl" with any of
  the following names:

    gslConfig.cmake
    gsl-config.cmake

  Add the installation prefix of "gsl" to CMAKE_PREFIX_PATH or set "gsl_DIR"
  to a directory containing one of the above files.  If "gsl" provides a
  separate development package or SDK, be sure it has been installed.
Call Stack (most recent call first):
  CMakeLists.txt:51 (fetch_gsl)


-- Configuring incomplete, errors occurred!
See also "/home/pavlis/src/mspass/cxx/build/CMakeFiles/CMakeOutput.log".
```
This was the error I got earlier I thought I could solve, but I couldn't find any useful reference the web that could give me a hint about where I could get the gslConfig.cmake file what should be in it.   

Curiously I also installed gsl with apt install.  That completed successfully, but after doing so the cmake process did not find the local system installation but went through the download and compile sequence again throwing the above error.  

So, one solution is to better understanding what directories find_package search and how they are configured.   Then I could use the local copy on this machine.   I now have a workaround that makes this less critical.   I can work on my laptop and do some initial debugging of the deconvolution stuff.   I need to next build a C++ test suite.   I had something for the antelope version, but it has dependencies on the stream io library that make it impossible to port.   Let me know if you can come up with a solution for this, but don't waste your weekend.   This can wait.

This problem does not seem to be limited to the version of cmake.  I installed much newer version on my ubuntu machine.   Further, after some confusion created by a system problem that apparently went away when the did system maintenance today, I get the same error on the carbonate machine.   This may be a trivial fix, but I'm stumped.   I really need to find some kind of more comprehensive document on cmake so I'm no just poking in the dark and resorting to google searches to fix problems like this.  Anyway, here is the message I get consistently on linux ubuntu and redhat (carbonate):

100%] Built target gsl
CMake Error at cmake/gsl.cmake:25 (find_package):
  By not providing "Findgsl.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "gsl", but
  CMake did not find one.

  Could not find a package configuration file provided by "gsl" with any of
  the following names:

    gslConfig.cmake
    gsl-config.cmake

  Add the installation prefix of "gsl" to CMAKE_PREFIX_PATH or set "gsl_DIR"
  to a directory containing one of the above files.  If "gsl" provides a
  separate development package or SDK, be sure it has been installed.
Call Stack (most recent call first):
  CMakeLists.txt:55 (fetch_gsl)


-- Configuring incomplete, errors occurred!
See also "/N/u/pavlis/Carbonate/src/mspass/cxx/build/CMakeFiles/CMakeOutput.log".

Note in both cases gsl is improperly built.  I had gsl installed on my ubuntu machine as a standard package. On carbonate I had their "modules" system set to know about gsl, but cmake is simply not finding it.  Then after it builds it's own copy it can't seem to configure itself.   Any ideas?

Do you have any idea how to silence this kind of bogus warning message:
```
    if(nsamp>d.size())
/home/pavlis/src/mspass/cxx/include/mspass/deconvolution/ComplexArray.h:180:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
```
Any time a C++ program uses the std::vector size method in a construct like that where a regular
int is compared to the output you cget the -Wsign-compare warning above.   That construct is very common and a somewhat silly warning.  It probably should only be present for if one of my favorite compiler flag is set:  -Wpedantic.  There is an obscure English word for you, but it totally describes this particular error.  Why comparison to a signed and unsigned int should generate a warning by default is beyond me since even basic programming books contain such constructs.   

Fixing this requires some global parameter change for cmake.  I think it is probably comes from having the CMAKE_FLAGS_DEBUG argument set to -Wall, but I'm reluctant to change that because it will suppress some other warning that can be useful.

Note this seems to only be an issue in newer versions of gcc.  Older revs do not behave this way.

--- a/cxx/cmake/yaml-cpp.cmake
+++ b/cxx/cmake/yaml-cpp.cmake
@@ -22,7 +22,6 @@ macro(fetch_yaml_cpp _download_module_path _download_root)
             ${_download_root}
         )
 #    list (APPEND CMAKE_MODULE_PATH ${PROJECT_BINARY_DIR}/yaml-cpp/lib/cmake/yaml-cpp/)
-    set (yaml-cpp_DIR ${PROJECT_BINARY_DIR}/yaml-cpp/lib/cmake/yaml
--cpp/)
+    set (yaml-cpp_DIR ${PROJECT_BINARY_DIR}/yaml-cpp/lib/cmake/yaml-cpp/)
     find_package (yaml-cpp REQUIRED)
 endmacro()

--- a/docs/source/getting_started/advanced_setup_considerations.rst
+++ b/docs/source/getting_started/advanced_setup_considerations.rst
@@ -17,6 +17,9 @@ the ``mspasspy.ccore`` C++ extension.  Follow the project's
 for the required compilers and libraries.  The current package metadata
 requires Python 3.10 or newer.
 
+Core maintainers adding or changing a compiled third-party library should
+also follow :ref:`cmake_dependency_maintenance`.
+
 Use an isolated environment
 ---------------------------
 

--- a/docs/source/getting_started/cmake_dependency_maintenance.rst
+++ b/docs/source/getting_started/cmake_dependency_maintenance.rst
@@ -11,7 +11,9 @@ The official CMake `Using Dependencies Guide
 <https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html>`__
 is the best general introduction.  MsPASS currently uses a system-first
 variant of that model: use an installed package when CMake can find it and,
-for selected dependencies, build a pinned source release otherwise.
+for selected dependencies, build from source otherwise.  Most fallbacks pin
+a release; the current GSL fallback's ``gsl-latest.tar.gz`` URL is a legacy
+exception and is not reproducible in the same way.
 
 How dependency discovery works
 ------------------------------
@@ -27,8 +29,9 @@ dependencies follow this sequence:
 #. The macro configures ``cxx/cmake/<package>-download.cmake`` as a small,
    separate CMake project.  That project uses `ExternalProject
    <https://cmake.org/cmake/help/latest/module/ExternalProject.html>`__ to
-   download, build, and install a pinned release beneath the MsPASS build
-   directory.
+   download, build, and install the configured source beneath the MsPASS
+   build directory.  New fallbacks should use a fixed release or commit;
+   GSL's moving ``gsl-latest.tar.gz`` URL is an existing exception.
 #. The macro sets the package-specific search hint, such as
    ``<Package>_DIR`` or ``<Package>_ROOT``, then calls ``find_package`` with
    ``REQUIRED``.  The rest of the build therefore consumes the same package
@@ -59,10 +62,11 @@ entry when that layer actually consumes or distributes the dependency.
    set the correct search hint, and finish with a required package lookup.
 
 ``cxx/cmake/<package>-download.cmake``
-   Define the pinned source URL or Git tag, checksum when an archive is used,
-   configure/build/install commands, and an install prefix inside the build
-   tree.  Static libraries linked into the Python extension must be compiled
-   as position-independent code; see CMake's
+   Pin the source URL, Git tag, or commit; provide a checksum when an archive
+   is used; and define the configure/build/install commands and an install
+   prefix inside the build tree.  Do not copy the legacy moving GSL URL when
+   adding a fallback.  Static libraries linked into the Python extension must
+   be compiled as position-independent code; see CMake's
    `POSITION_INDEPENDENT_CODE
    <https://cmake.org/cmake/help/latest/prop_tgt/POSITION_INDEPENDENT_CODE.html>`__
    reference.

--- a/docs/source/getting_started/cmake_dependency_maintenance.rst
+++ b/docs/source/getting_started/cmake_dependency_maintenance.rst
@@ -1,0 +1,146 @@
+.. _cmake_dependency_maintenance:
+
+Maintaining C++ Dependencies
+============================
+
+This page is for core contributors who add or change a compiled dependency.
+End users building MsPASS should follow the :ref:`source-build instructions
+<advanced_setup_considerations>` instead.
+
+The official CMake `Using Dependencies Guide
+<https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html>`__
+is the best general introduction.  MsPASS currently uses a system-first
+variant of that model: use an installed package when CMake can find it and,
+for selected dependencies, build a pinned source release otherwise.
+
+How dependency discovery works
+------------------------------
+
+``cxx/CMakeLists.txt`` is the entry point for the compiled build.  Most
+dependencies follow this sequence:
+
+#. Call `find_package
+   <https://cmake.org/cmake/help/latest/command/find_package.html>`__ without
+   ``REQUIRED`` so an installed package is preferred.
+#. If the package was not found, include ``cxx/cmake/<package>.cmake`` and
+   call its ``fetch_<package>`` macro.
+#. The macro configures ``cxx/cmake/<package>-download.cmake`` as a small,
+   separate CMake project.  That project uses `ExternalProject
+   <https://cmake.org/cmake/help/latest/module/ExternalProject.html>`__ to
+   download, build, and install a pinned release beneath the MsPASS build
+   directory.
+#. The macro sets the package-specific search hint, such as
+   ``<Package>_DIR`` or ``<Package>_ROOT``, then calls ``find_package`` with
+   ``REQUIRED``.  The rest of the build therefore consumes the same package
+   variables or imported targets for either discovery path.
+#. The library and header information is attached to the targets that use the
+   dependency.  ``cxx/src/lib/CMakeLists.txt`` links the combined ``mspass``
+   library; component and Python-extension ``CMakeLists.txt`` files add any
+   direct include or link requirements.
+
+The current tree contains concrete examples for yaml-cpp, Boost, GSL,
+OpenBLAS, libmseed, and pybind11 under ``cxx/cmake``.  yaml-cpp, Boost, GSL,
+BLAS/LAPACK, and pybind11 first try an installed package.  libmseed is always
+built from its pinned source release.  Python itself must already be present.
+
+Files that normally change
+--------------------------
+
+Adding a dependency requires checking each layer below.  Only add a file or
+entry when that layer actually consumes or distributes the dependency.
+
+``cxx/CMakeLists.txt``
+   Add discovery, the optional source-build fallback, and status output that
+   identifies the resolved library and include locations.
+
+``cxx/cmake/<package>.cmake``
+   Add this orchestration macro only when MsPASS will build the package if it
+   is absent.  It should configure and build the companion download project,
+   set the correct search hint, and finish with a required package lookup.
+
+``cxx/cmake/<package>-download.cmake``
+   Define the pinned source URL or Git tag, checksum when an archive is used,
+   configure/build/install commands, and an install prefix inside the build
+   tree.  Static libraries linked into the Python extension must be compiled
+   as position-independent code; see CMake's
+   `POSITION_INDEPENDENT_CODE
+   <https://cmake.org/cmake/help/latest/prop_tgt/POSITION_INDEPENDENT_CODE.html>`__
+   reference.
+
+Consumer ``CMakeLists.txt`` files
+   Add the dependency to the narrowest target that needs it.  Prefer an
+   imported target supplied by the package when available; otherwise use the
+   include and library variables exposed by its Find module or config file.
+   See `target_link_libraries
+   <https://cmake.org/cmake/help/latest/command/target_link_libraries.html>`__.
+
+Package and container manifests
+   Keep ``pyproject.toml``, ``scripts/dependency_map.toml``, ``meta.yaml``,
+   ``scripts/conda_build.sh``, and the relevant Docker image stages aligned.
+   A dependency built only as a private static fallback may not need a runtime
+   package entry, but a shared library required after installation does.
+
+CI and tests
+   Update ``.github/workflows/cmake.yml`` or a focused test when the existing
+   build does not exercise the new discovery or fallback behavior.
+
+Package search hints
+--------------------
+
+Do not create an MsPASS package-config file merely to make ``find_package``
+work.  CMake may use either a bundled ``Find<Package>.cmake`` module or a
+``<Package>Config.cmake``/``<package>-config.cmake`` file installed by the
+dependency.
+
+For a dependency installed outside standard system prefixes, configure with
+``CMAKE_PREFIX_PATH`` set to its installation prefix.  For config-mode
+packages, ``<Package>_DIR`` must name the directory containing the package's
+config file, not the library file or the prefix above it.  Package names and
+their result variables are case-sensitive; use the spelling documented by
+the corresponding Find module or upstream package.
+
+Validation checklist
+--------------------
+
+Before opening a pull request for a compiled dependency change:
+
+#. Configure and build with the dependency installed by the system or active
+   environment.
+#. In a clean container or environment where it is absent, configure again
+   and verify that the intended fallback is selected and installed beneath
+   the build directory.  Do not uninstall packages from a working developer
+   machine just to exercise this path.
+#. Build the Python extensions and run CTest:
+
+   .. code-block:: bash
+
+      cmake -S cxx -B /tmp/mspass-cmake-build -DCMAKE_BUILD_TYPE=Release
+      cmake --build /tmp/mspass-cmake-build --parallel
+      ctest --test-dir /tmp/mspass-cmake-build --output-on-failure
+
+#. If package manifests changed, verify their generated copies:
+
+   .. code-block:: bash
+
+      python3 scripts/sync_dependencies.py --check
+
+#. Confirm both configure logs identify the expected headers and libraries.
+   A successful download alone is insufficient; the final required
+   ``find_package`` call and the consuming target must resolve the built
+   package.
+
+Common failures
+---------------
+
+* A static archive used by a Python extension was not compiled with ``-fPIC``
+  or ``CMAKE_POSITION_INDEPENDENT_CODE=ON``.
+* ``<Package>_DIR`` points at an installation prefix instead of the directory
+  containing the config file.
+* The package name has the wrong case, so CMake selects a different search
+  mode or fails to load its Find module.
+* Headers were added globally but the library was not linked to the consuming
+  target, or the inverse.
+* A Git branch was used without a fixed tag or commit, making builds change
+  over time.
+* Only the installed-package path was tested, leaving a broken fallback that
+  CI never exercises.

--- a/docs/source/index.rst
+++ b/docs/source/index.rst
@@ -125,6 +125,7 @@ Choose your path
 
    user_manual/FAQ
    user_manual/development_strategies
+   CMake dependency maintenance <getting_started/cmake_dependency_maintenance>
 
 .. toctree::
    :maxdepth: 2

